lib/collection/src/collection.rs

```diff
@@ -642,38 +643,131 @@ impl Collection {
+                        req.with_vector,
+                        shard_selection,
+                    )
+                });
+            try_join_all(filled_results).await
+        } else {
+            let result = self
+                ._search_batch(request, search_runtime_handle, shard_selection)
+                .await?;
+            Ok(result)
+        }
+    }
+
+    pub async fn _search_batch(
+        &self,
+        request: SearchRequestBatch,
+        search_runtime_handle: &Handle,
+        shard_selection: Option<ShardId>,
+    ) -> CollectionResult<Vec<Vec<ScoredPoint>>> {
         let request = Arc::new(request);
 
-        let all_searches_res = {
+        // query all shards concurrently
+        let mut all_searches_res = {
             let shard_holder = self.shards_holder.read().await;
             let target_shards = shard_holder.target_shards(shard_selection)?;
             let all_searches = target_shards
                 .iter()
                 .map(|shard| shard.get().search(request.clone(), search_runtime_handle));
-            try_join_all(all_searches).await?.into_iter().flatten()
+            try_join_all(all_searches).await?
         };
 
-        let distance = self.config.read().await.params.distance;
-        let mut top_result = match distance.distance_order() {
-            Order::LargeBetter => {
-                peek_top_largest_scores_iterable(all_searches_res, request.limit + request.offset)
-            }
-            Order::SmallBetter => {
-                peek_top_smallest_scores_iterable(all_searches_res, request.limit + request.offset)
+        // merge results from shards in order
+        let mut merged_results: Vec<Vec<ScoredPoint>> = vec![vec![]; request.searches.len()];
+        for shard_searches_results in all_searches_res.iter_mut() {
+            for (index, shard_searches_result) in shard_searches_results.iter_mut().enumerate() {
+                merged_results[index].append(shard_searches_result)
             }
-        };
-
-        if request.offset > 0 {
-            // Remove offset from top result.
-            top_result.drain(..request.offset);
         }
-        Ok(top_result)
+        let distance = self.config.read().await.params.distance;
+        let top_results: Vec<Vec<ScoredPoint>> = merged_results
+            .into_iter()
+            .zip(request.searches.iter())
+            .map(|(res, request)| {
+                let mut top_res = match distance.distance_order() {
+                    Order::LargeBetter => {
+                        peek_top_largest_scores_iterable(res, request.limit + request.offset)
+                    }
+                    Order::SmallBetter => {
+                        peek_top_smallest_scores_iterable(res, request.limit + request.offset)
+                    }
+                };
+                // Remove offset from top result.
+                if request.offset > 0 {
+                    if top_res.len() >= request.offset {
```

Comment:
Added a fix after testing the usage of offsets.
The bug must be present in our regular search already.